### ISSUE TITLE
Recommendations for a non-realistic character body

### ISSUE THREAD

**Issue by Kazade** [irlanrobson/bounce#46]
I'm looking to implement something similar to Unity's Character Controller. There's a good description of it here: https://docs.unity3d.com/Manual/class-CharacterController.html

What would be the recommended way to do something similar in Bounce (i.e. collidable, but no traditional acceleration etc.)?

**Comment by irlanrobson** [comment]
Hi Kazade.

There is a great portion of presentation introducing this topic [here](http://box2d.org/files/GDC2014/GDC2014_ErinCatto.zip
) at the end. 

I've never implemented a character controller. The algorithm seems not difficult to digest and implement since there aren't any rotations involved, you only translate the capsule shape. There might be edge cases though. I can add a shape cast function to the library if you are interested in implementing the position solver described in this resource.

**Comment by Kazade** [comment]
That's a really useful link. Thanks!

I'll have a read but might just have to rush something for now with what's currently in Bounce. I imagine a ShapeCast function would be useful anyway though!

**Comment by irlanrobson** [comment]
The shape cast function has been added. It's called b3GJKRayCast and it's located inside bounce/collision/gjk.h. 

There's also a nice test namely ShapeCast which uses this function. Maybe you can look that up for example.

**Comment by Kazade** [comment]
Thanks! You're too awesome!

**Comment by Kazade** [comment]
I've been doing some more research on this, and I've found an interesting approach documented on this blog: https://roystanross.wordpress.com/2014/05/07/custom-character-controller-in-unity-part-1-collision-resolution/

The approach they take is, rather than shape-casting to see if the character is about to collide with something, they wait until they *are* colliding and move the character to the closest point on the colliding shape.

The obviously difficult part is finding the closest point on the surface of a shape...

**Comment by irlanrobson** [comment]
I quick scanned through the articles and it looks it's trying to accomplish what Catto's algorithm does. 

Here's how Catto's algorithm I recommended works in practice. You can test this using a simple sphere as the character shape or even a point.

Initialization:

1. Check if the character shape overlaps with each static shape in the environment. For each overlap you build a plane using the contact point and normal and add this plane to the plane list.

2. Using the plane list you solve the target point to get the current solve point.

That's the looped part:

3. Sweep from the start point to the solve point. For each sweep hit you build a plane using the point and normal at the time of impact. You add this plane to the plane list.

4. Using the plane list you solve the target point to get solve point.
 
3. Terminate if the old solve point and solve point are close enough.

Here's a pseudo-code. You can easely see how it works if you sketch this in a paper. Note that you must not clear the plane list. 


```

    void SolveCharacter()
	{
		// Start point
		b3Vec3 start_point = GetCharacterPosition();

		// Target position
		b3Vec3 target_point = GetTargetPosition();

		// Clip planes
		b3Array<b3Plane> planes;
		
		// Collect overlap planes
		CollectOverlapPlanes(planes);
		
		// Plane solve to get current solve point
		b3Vec3 solve_point = PlaneSolve(target_point, planes);

		// Old solve point
		b3Vec3 old_solve_point = solve_point;
		
		for (;;)
		{
			// Vector from start point to current solve point. This is the sweep path.
			b3Vec3 translation = solve_point - start_point;
			
			// Collect sweep planes. 
			CollectSweepPlanes(planes, translation);

			// Plane solve
			solve_point = PlaneSolve(target_point, planes);

			const float32 tolerance = 0.2f;
			// Terminate
			if (b3DistanceSquared(old_solve_point, solve_point) < tolerance * tolerance)
			{
				break;
			}

			old_solve_point = solve_point;
		}
		
		// Now solve point is the valid character position.
	}	
	
	b3Vec3 PlaneSolve(const b3Vec3& target_point, const b3Array<b3Plane>& planes)
	{
		b3Vec3 p = target_point;
		
		const u32 kMaxPositionIterations = 20;
		
		// Solve position constraints
		for (u32 i = 0; i < kMaxPositionIterations; ++i)
		{
			for (u32 j = 0; j < planes.Count(); ++j)
			{
				b3Plane plane = planes[j].plane;
				float32 s = b3Distance(p, plane);

				if (s > 0.0f)
				{
					continue;
				}

				b3Vec3 n = plane.normal;

				p -= s * n;
			}
		}
		
		return p;
	}
	
	void CollectOverlapPlanes(b3Array<b3Plane>& planes)
	{
		    Sphere sphere2 = GetCharacterSphere();
			
			// For each static shape in the static environment...
			for (u32 i = 0; i < shapes.Count(); ++i)
			{
				Shape shape1 = shapes[i];
				
				OverlapContact contact;
				bool overlap = Overlap(&contact, shape1, sphere2);

				if (overlap)
				{
					b3Plane plane(contact.normal, contact.point);
					
					planes.PushBack(plane);
				}
			}
	}
	
	void CollectSweepPlanes(b3Array<b3Plane>& planes, const b3Vec3& translation)
	{
		    Sphere sphere2 = GetCharacterSphere();
			
			// For each static shape in the static environment...
			for (u32 i = 0; i < shapes.Count(); ++i)
			{
				Shape shape1 = shapes[i];
				
				SweepContact contact;
				bool hit = Sweep(&contact, shape1, sphere2, translation2);

				if (hit)
				{
					b3Plane plane(contact.normal, contact.point);
					
					planes.PushBack(plane);
				}
			}
	}

```

**Comment by Kazade** [comment]
That's really helpful! Thank you! Just a few small questions:

 - Just to clarify, should the player be a static body (as it's not part of the simulation)?
 - Would I use b3GJKRayCast for both `Sweep()` and `Overlap()`?

**Comment by irlanrobson** [comment]
That could be a kinematic body. 

The function Sweep within the pseudo-code above was a short for b3GKJRayCast.

The function Overlap is contact creation. I meant you should grab a contact point, normal, and separation if the collision between the character sphere and the other shape has been detected. In your test system you can use b3Shape::TestSphere. It takes a sphere and the shape transform as input and outputs the contact point, normal, and separation if the sphere is overlapping that shape. Overlap is called only once for initializing the array of planes.

One thing I should mention is that b3GJKRayCast is not optimal and can fail in certain situations. In that case you would increase the maximum number of iterations defined in this function, namely kMaxIters. The file is gjk.cpp.

**Comment by irlanrobson** [comment]
I've just renamed the function b3GJKRayCast to b3GJKShapeCast for your convenience.

**Comment by irlanrobson** [comment]
I've added a test called CharacterTest to Testbed. There is a simple character controller which works and can be developed. 

Hope that helps.

**Comment by Kazade** [comment]
Thank you!

**Comment by Lagger625** [comment]
@irlanrobson what happened to the CharacterTest?

**Comment by irlanrobson** [comment]
It was removed because it was just a draft. I don't want any drafts in this project. 

You can find it through the history if your GitHub Fu is better than mine. :)
